File: Override/2.py
class Cal(object):
  _history = []
  def __init__(self, v1, v2): #self는 변수 이름. 다른 것도 가능
    if isinstance(v1, int):
      self.v1 = v1
    if isinstance(v2, int):
      self.v2 = v2
  def add(self):
    result = self.v1+self.v2
    Cal._history.append(f"{self.v1}+{self.v2}={result}")
    return result
  def subtract(self):
    result = self.v1-self.v2
    Cal._history.append(f"{self.v1}-{self.v2}={result}")
    return result
  def setV1(self, v):
    if isinstance(v, int):
      self.v1 = v
  def getV1(self):
    return self.v1
  @classmethod
  def history(cls):
    for i in Cal._history:
      print(i)
  def info(self):
    return "Cal => v1 : %d, v2 : %d" % (self.v1, self.v2)
class Calmultiply(Cal): #Cal의 모든 메소드 상속받음
  def multiply(self):
    result = self.v1*self.v2
    Cal._history.append(f"{self.v1}*{self.v2}={result}")
    return result
  def info(self):
    return f"CalMultiply => {super().info()}"
class Caldivide(Calmultiply):
  def divide(self):
    result = self.v1/self.v2
    Cal._history.append(f"{self.v1}/{self.v2}={result}")
    return result
  def info(self):
    return f"Caldivide => {super().info()}"
c0 = Cal(30,60)
print(c0.info())
c1 = Calmultiply(10,10)
print(c1.info())
c2 = Caldivide(20,10)
print(c2.info())

File: Override/2.rb
class Cal
  attr_accessor :v1 # reader, writer
  attr_accessor :v2
  @@_history = []
  def initialize(v1,v2)
    @v1,@v2 = v1,v2
  end
  def Cal.history()
    for item in @@_history
      p item
    end
  end
  def add()
    result = @v1+@v2
    @@_history.push("#{@v1}+#{@v2}=#{result}")
    return result
  end
  def subtract()
    result = @v1-@v2
    @@_history.push("#{@v1}-#{@v2}=#{result}")
    return result
  end
  def setV1(v)
    if v.is_a?(Integer)
      @v1 = v
    end
  end
  def getV1()
    return @v1
  end
  def info
    return "Cal => v1 : #{@v1}, v2 : #{@v2}"
  end
end

class Calmultiply < Cal
  def multiply()
    result = @v1*@v2
    @@_history.push("#{@v1}*#{@v2}=#{result}")
    return result
  end
  def info
    return "Calmultiply => #{super()}"
  end
end

class Caldivide < Calmultiply
  def divide()
    result = @v1/@v2
    @@_history.push("#{@v1}/#{@v2}=#{result}")
    return result
  end
  def info
    return "Caldivide => #{super()}"
  end
end
c0 = Cal.new(30,60)
p c0.info
c1 = Calmultiply.new(10,10)
p c1.info
c2 = Caldivide.new(20,10)
p c2.info